python/datasources/cdc_restricted_local.py

```diff
@@ -0,0 +1,251 @@
+                    values_mapping):
+    """Converts/adds columns for cases, hospitalizations, deaths. Does some
+    basic standardization of dataframe elements. Groups by given groupby_cols
+    and aggregates. Returns sum of the aggregated df & overall_df.
+
+    df: Pandas dataframe that contains a chunk of all of the raw data.
+    groupby_cols: List of columns we want to groupby / aggregate on.
+    overall_df: Pandas dataframe to add our aggregated data to.
+    demog_col: Name of the demographic column to standardize.
+    values_mapping: Mapping from demog value to standardized form.
+    """
+    # Add a columns of all ones, for counting the # of cases / records.
+    df[std_col.COVID_CASES] = np.ones(df.shape[0], dtype=int)
+
+    # Add columns for hospitalization yes/no/unknown and death yes/no/unknown,
+    # as we aggregate and count these individually. Do a sanity check that we
+    # covered all the data and drop the original hospitalization/death columns.
+    df[std_col.COVID_HOSP_Y] = (df['hosp_yn'] == 'Yes')
+    df[std_col.COVID_HOSP_N] = (df['hosp_yn'] == 'No')
+    df[std_col.COVID_HOSP_UNKNOWN] = ((df['hosp_yn'] == 'Unknown') |
+                                      (df['hosp_yn'] == 'Missing') |
+                                      (df['hosp_yn'] == 'OTH'))
+    df[std_col.COVID_DEATH_Y] = (df['death_yn'] == 'Yes')
+    df[std_col.COVID_DEATH_N] = (df['death_yn'] == 'No')
+    df[std_col.COVID_DEATH_UNKNOWN] = ((df['death_yn'] == 'Unknown') |
+                                       (df['death_yn'] == 'Missing'))
+
+    assert (df[std_col.COVID_HOSP_Y] | df[std_col.COVID_HOSP_N] |
+            df[std_col.COVID_HOSP_UNKNOWN]).all()
+    assert (df[std_col.COVID_DEATH_Y] | df[std_col.COVID_DEATH_N] |
+            df[std_col.COVID_DEATH_UNKNOWN]).all()
+
+    df = df.drop(columns='hosp_yn')
+    df = df.drop(columns='death_yn')
+
+    # Standardize the values in demog_col using values_mapping.
+    df = df.replace({demog_col: values_mapping})
+
+    # Group by the desired columns and compute the sum/counts of
+    # cases/hospitalizations/deaths. Add this df to overall_df.
+    df = df.groupby(groupby_cols).sum().reset_index()
+    df = df.set_index(groupby_cols)
+    if not overall_df.empty:
+        return overall_df.add(df, fill_value=0)
+    return df
+
+
+def sanity_check_data(df):
+    # Perform some simple sanity checks that we are covering all the data.
+    cases = df[std_col.COVID_CASES]
+    assert cases.equals(df[std_col.COVID_HOSP_Y] + df[std_col.COVID_HOSP_N] +
+                        df[std_col.COVID_HOSP_UNKNOWN])
+    assert cases.equals(df[std_col.COVID_DEATH_Y] + df[std_col.COVID_DEATH_N] +
+                        df[std_col.COVID_DEATH_UNKNOWN])
+
+
+def standardize_data(df):
+    """Standardizes the data by cleaning string values and standardizing column
+    names.
+
+    df: Pandas dataframe to standardize.
+    """
+    # Clean string values in the dataframe.
+    def _clean_str(x):
```

Comment:
Is there a way in Python for extension methods?  Like can this be added to the string python proto or something rather than a local func?

Reply:
this one actually i had an extra "strip" in there so i removed that and made it into an inline lambda